### PR TITLE
[refactor+new features] A generic channel manager

### PR DESCRIPTION
**Pull request by LipuFei**
This is a subset of PR #1515. After this one is merged, I will make a PR to make metadata working on GUI and make a new unit test for metadata using the RSS feed approach, which is less crazier than #1515.

1. This PR makes a generic `channel_manager` that handles the functionality related to the channels owned by myself, including:
  * creating new channels
  * modifying rss feeds attached to a channel owned by myself
  * rss feed parser that can parse and retrieve metadata
2. Certain old channel functionalities have been replaced with this new `channel_manager`.

### CODE REVIEW

**Comment by LipuFei** [comment]
retest this please

**Comment by LipuFei** [comment]
I need to fix this `ListBody` `PyDeadObject` problem now. It's killing me.

**Comment by LipuFei** [comment]
retest this please

**Comment by LipuFei** [comment]
@NielsZeilemaker @whirm The `PyDeadObject` problem should be fix. Ready for review.

File: Tribler/Core/Modules/channel/channel.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    def shutdown(self):
+        self.cancel_all_pending_tasks()
+        for key, rss_parser in self._rss_feed_dict.iteritems():
+            if rss_parser is not None:
+                rss_parser.shutdown()
+        self._rss_feed_dict = None
+        self._channel_community = None
+        self._session = None
+
+    @call_on_reactor_thread
```
Comment (NielsZeilemaker) [inline Tribler/Core/Modules/channel/channel.py]: Why are these methods run on the reactor thread?

File: Tribler/Core/Modules/channel/channel_manager.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        from Tribler.community.channel.community import ChannelCommunity
+        for community in self.session.lm.dispersy.get_communities():
+            if isinstance(community, ChannelCommunity) and community.master_member and community.master_member.private_key:
+                channel_obj = ChannelObject(self.session, community, is_created=True)
+                channel_obj.initialize()
+                self._channel_list.append(channel_obj)
+
+                self._logger.debug(u"loaded channel '%s', %s", channel_obj.name, hexlify(community.cid))
+
+    @blocking_call_on_reactor_thread
```
Comment (NielsZeilemaker) [inline Tribler/Core/Modules/channel/channel_manager.py]: Hmm, is this a good idea?

File: Tribler/Main/metadata-injector.py [hunk 3]
```diff
@@ -31,21 +32,31 @@
 from Tribler.dispersy.util import call_on_reactor_thread
 
 
-def parge_rss_config_file(file_path):
-    rss_list = []
+def parge_channel_config_file(file_path):
```
Comment (NielsZeilemaker) [inline Tribler/Main/metadata-injector.py]: parse?

File: Tribler/Main/vwxGUI/list_body.py [hunk 4]
```diff
@@ -763,6 +763,9 @@ def OnCollapse(self, item=None, raise_events=True, from_expand=False):
 
     @warnWxThread
     def OnChange(self, scrollToTop=False):
+        if GUIUtility.getInstance().utility.abcquitting:
```
Comment (NielsZeilemaker) [inline Tribler/Main/vwxGUI/list_body.py]: Why is this here?

File: Tribler/Core/Modules/channel/channel.py (hunk 1 above)
Comment (LipuFei) [inline Tribler/Core/Modules/channel/channel.py]: @NielsZeilemaker Here we have two functions that can change `rss_feed_dict`: one is create_rss_feed(), which is an API function directly called by others, and the other is this `_on_channel_created()`, which is called by the `Notifier`. I want to make sure that only one guy can do something with the `rss_feed_dict`, so I made them both `call_on_reactor_thread`.

File: Tribler/Core/Modules/channel/channel_manager.py (hunk 2 above)
Comment (LipuFei) [inline Tribler/Core/Modules/channel/channel_manager.py]: @NielsZeilemaker There is no other threads involved here except the reactor thread, so it is safe to do this here.

File: Tribler/Main/metadata-injector.py (hunk 3 above)
Comment (LipuFei) [inline Tribler/Main/metadata-injector.py]: @NielsZeilemaker Nice one! I will fix it.

File: Tribler/Core/Modules/channel/channel.py (hunk 1 above)
Comment (NielsZeilemaker) [inline Tribler/Core/Modules/channel/channel.py]: Put a lock around them, there is no need to push them on the reactor thread.

File: Tribler/Core/Modules/channel/channel_manager.py (hunk 2 above)
Comment (NielsZeilemaker) [inline Tribler/Core/Modules/channel/channel_manager.py]: Is there a need for it?

File: Tribler/Main/vwxGUI/list_body.py (hunk 4 above)
Comment (LipuFei) [inline Tribler/Main/vwxGUI/list_body.py]: This is to fix the mysterious `PyDeadObject` problem in `ListBody`. See the first few test results for the backtrace.

This problem happens during the shutdown.  I tracked it down to this `SetupScrolling()`. This is a wx function which schedules `_SetupAfter()` with `wx.CallAfter()`, and this `_SetupAfter()` is the one that raised the `PyDeadObject` error.

The only place I found that calls `SetupScrolling()` is this `OnChannge()` in `ListBody`. Because this new RSS parser I made is too fast (compared to the old one) that in the rss_import test, you can see tens of torrents showing up simultaneously on the "Channel Manage" page, I suspect that this performance improvement might be the issue.

In short, the solution to fix this `PyDeadObject` problem is to not let `ListBody` update itself during the shutdown. So I added this line.

Comment (NielsZeilemaker) [inline Tribler/Main/vwxGUI/list_body.py]: List_body should be a generic component, no need for a call to guiutility here.
A PyDeadObject can be fixed by doing
```
if self:
```
no need for abcquitting

Comment (LipuFei) [inline Tribler/Main/vwxGUI/list_body.py]: This is the tricky part. `if self` doesn't work for this problem. The issue is that when this `OnChange()` is called by an wx event and `ListBody` is not dead yet. So `_SetupAfter()` has scheduled but somehow not executed before `ListBody` is dead. After we exit the MainLoop and wake up the MainThread again (for the shutdown splash window), this `_SetupAfter()` got scheduled and `PyDeadObject` appears.

So the only way I found is to use this `GUIUtility.getInstance().utility.abcquitting` to avoid calling `SetupScrolling()`.

File: Tribler/Core/Modules/channel/channel_manager.py (hunk 2 above)
Comment (LipuFei) [inline Tribler/Core/Modules/channel/channel_manager.py]: I will make this one a `call_on_reactor_thread`.

File: Tribler/Core/Modules/channel/channel.py (hunk 1 above)
Comment (LipuFei) [inline Tribler/Core/Modules/channel/channel.py]: I will schedule a task with `self.register_task()`. This way, we can cancel the tasks without having them dangling around.

**Comment by LipuFei** [comment]
@NielsZeilemaker @whirm ready

**Comment by LipuFei** [comment]
retest this please

File: Tribler/community/channel/community.py [hunk 5]
```diff
@@ -726,21 +728,23 @@ def _disp_check_modification(self, messages):
                     if infohash:
                         infohash = str2bin(infohash)
                         logger.debug(
-                            "Incoming swift-thumbnails with infohash %s from %s",
+                            "Incoming metadata-json with infohash %s from %s",
                             infohash.encode("HEX"),
                             message.candidate.sock_addr[0])
 
-                        if not th_handler.has_metadata("thumbs", infohash, thumbnail_subpath):
+                        print "Incoming metadata-json with infohash %s from %s" % (infohash.encode("HEX"),
```
Comment (NielsZeilemaker) [inline Tribler/community/channel/community.py]: I guess you should remove this line, otherwise. It's ready right @LipuFei ?

Comment (LipuFei) [inline Tribler/community/channel/community.py]: It is. I will remove this line first

**Comment by NielsZeilemaker** [comment]
Go ahead and merge this one. What issues are you guys working on btw?

**Comment by LipuFei** [comment]
@whirm Ready

@NielsZeilemaker Not Many. We have some issues that @synctext needs to check.
#1568 is tricky; I cannot reproduce #1569;
Now I'm working on #1574 and #1575

**Comment by NielsZeilemaker** [comment]
I would create a unit-test for #1568, a special build for #1569 which enables the debug log of the quit methods in mainframe.py. Also #1540 seems like a trivial fix.

File: Tribler/Main/vwxGUI/channel.py [hunk 6]
```diff
@@ -1519,20 +1521,26 @@ def OnChange(self, event):
...
         item = event.GetEventObject()
 
-        self.torrentfeed.deleteURL(item.url, self.channel.id)
-        self.RebuildRssPanel()
+        my_channel_object = self.guiutility.utility.session.lm.channel_manager.get_my_channel(self.channel.id)
+        my_channel_object.remove_rss_feed(item.url)
 
     def OnRefreshRss(self, event):
-        self.torrentfeed.doRefresh()
+        # TODO(lipu): fix this
```
Comment (whirm) [inline Tribler/Main/vwxGUI/channel.py]: Forgot this?

**Comment by whirm** [comment]
The rest seems fine

File: Tribler/Main/vwxGUI/channel.py (hunk 6 above)
Comment (LipuFei) [inline Tribler/Main/vwxGUI/channel.py]: This will be fixed by the next PR. I have split this metadata thing into several smaller parts. It's a remainder for myself..

**Comment by LipuFei** [comment]
@whirm That TODO should go actually. I removed it and fixed one more typo. Ready when it passes.

**Comment by whirm** [comment]
Great
